plugins/roulette/roulette.7
.\"
.\" Copyright (c) 2013-2026 David Demelier <markand@malikania.fr>
.\"
.\" Permission to use, copy, modify, and/or distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd @IRCCD_MAN_DATE@
.Dt IRCCD-PLUGIN-ROULETTE 7
.Os
.\" NAME
.Sh NAME
.Nm roulette
.Nd irccd roulette plugin
.\" DESCRIPTION
.Sh DESCRIPTION
The
.Nm
plugin is a funny script that let you do a russian roulette game but without any
injuries.
.Pp
Note: the bot will perform a kick command if the user is being shot so you must
be sure to have channel operator mode set on irccd.
.\" INSTALLATION
.Sh INSTALLATION
The plugin
.Nm
is distributed with irccd. To enable it add it into your
.Xr irccd.conf 5
configuration file.
.Pp
.Bd -literal -offset indent
plugin "roulette"
.Ed
.\" USAGE
.Sh USAGE
The
.Nm
plugin simply reacts to the special command.
.Bd -literal -offset Ds
markand: !roulette
irccd: markand, you're lucky this time
markand: !roulette
irccd: markand, you're lucky this time
markand: !roulette
markand was kicked by irccd [markand, HEADSHOT]
.Ed
.\" TEMPLATES
.Sh TEMPLATES
The plugin
.Nm
supports the following templates properties:
.Bl -tag -width lucky
.It Va lucky
The text to show on luck. Keywords:
.Em count
(the number of turns left in the cylinder).
.It Va shot
The text to show on shot.
.El
.Pp
All templates supports the following standard keywords:
.Em channel , command , nickname , origin , plugin , server .
.\" EXAMPLES
.Sh EXAMPLES
Example of configuration file:
.Bd -literal
plugin "roulette" {
  templates {
    "lucky" "#{nickname} you're gonna get shot";
    "shot"  "BIM";
  }
}
.Ed
.\" SEE ALSO
.Sh SEE ALSO
.Xr irccd 1 ,
.Xr irccd-templates 7


CHANGES.md
IRC Client Daemon CHANGES
=========================

irccd 5.0.0 ????-??-??
======================

This is a major release. See MIGRATING.md file for more information.

The key highlight of this release is the integration of libev as the main loop
and coroutines, making the code simpler and more expressive.

Build system
------------

The build system has been changed from CMake to GNU Make, this is because CMake
makes simple things complicated.

C API
-----

- Most of the string based types are dynamically allocated rather than creating
  dozens of arbitrary fixed size limits.
- Functions for allocated types are now coupled as `foo_new` and `foo_free` or
  `foo_init` and `foo_finish` otherwise.
- Types that require string list are now implemented using NULL sentinel based
  lists.

irccd
-----

- The auto-reconnect feature is automatically enabled on a server and is no
  longer an option. As an IRC bot, it make sense that a server should be
  reconnected permanently.
- All deferred functions have been removed.
- The server connection procedure is now implemented using coroutines for a
  more readable code logic.

misc
----

The C and Javascript APIs are now documented using doxygen rather than manual
pages.

irccd 4.1.1 2026-01-07
======================

- Fix minor issue in Irccd.Http API when `curl_multi_fdset` can return `-1` file
  descriptor.

irccd 4.1.0 2025-04-23
======================

- Add `irc_bot_pollable_add` to insert a custom interface into the main irccd
  loop.
- A new option `IRCCD_WITH_HTTP` has been added to accommodate the new curl
  dependency for `Irccd.Http` API.
- The plugin `history` manages two different timestamps for message and
  visibility. This helps showing the last message date of someone who is present
  on a channel but idle.

javascript API
--------------

- Brand new asynchronous `Irccd.Http` API.

plugins
-------

- links: rewritten in Javascript using the brand new `Irccd.Http` API.

irccd 4.0.4 2023-08-03
======================

- Improve tests running on a slow machines.
- Fix a warning on a invalid comparison.

irccd 4.0.3 2023-02-08
======================

- Fix build on NetBSD 9.3.

irccd 4.0.2 2023-01-18
======================

- Fix crash because of undefined variable.

irccd 4.0.1 2022-02-07
======================

- Fix disconnection with multiple servers.
- Fix Javascript `Irccd.Server` constructor function.
- Substitute the irccd version in Javascript plugins.
- Fix `irccdctl plugin-reload` command without argument.

irccd 4.0.0 2022-02-03
======================

This is a major release. See MIGRATING.md file for more information.

The key highlight for this release is the rewrite from C++ to C. The only
runtime dependency required is OpenSSL (if built with SSL support).

irccd
-----

- Irccd keeps track of nicknames in channels by capturing join/part/kick and
  mode changes. It is now more convenient from the plugins to quickly inspect if
  someone is present on a channel.
- It is now possible to change uid/gid of the transport socket file. The file is
  also created with permissions 664.
- A new `paths` command has been added to show default paths.

irccdctl
--------

- Commands `plugin-reload` and `plugin-unload` can be invoked without arguments.
- New `plugin-template` and `plugin-path` command which are synonyms of
  `plugin-config` but for templates and paths respectively.

plugins
-------

- tictactoe: now has a timeout in case of inactivity.

misc
----

- Split irccd-api manual page into individual irccd-api-<module> for a better
  readability.
- New `irccd.conf` and `irccdctl.conf` syntax.

network API
-----------

- Network protocol uses plain text again.
- Transport uses clear UNIX sockets only without passwords.

javascript API
--------------

- Brand new Irccd.Rule API to inspect and manage rules.
- Brand new Irccd.Hook API to inspect and manage hooks.

irccd 3.1.1 2021-01-04
======================

- Synchronize `ping-timeout` option in `[server]` to 1800 seconds by default.
- Enable `auto-reconnect` option in `[server]` by default as specified in the
  manual page.

irccd 3.1.0 2020-07-03
======================

- Added a new hook system. Hooks consist of an alternative approach to plugins
  to extend irccd in any language.

irccd 3.0.3 2019-10-06
======================

- Fix errors in irccdctl.conf example file.
- Add example of password in irccdctl.conf and irccd.conf.

irccd 3.0.2 2019-09-22
======================

- Added *IRCCD_WITH_JS* CMake variable in irccd package.
- Fixed trailing CTCP escape code.
- Fixed invalid parameters in topic event.
- Fixed invalid documentation of Irccd.System.usleep function.
- Fixed invalid system configuration directory.

irccd 3.0.1 2019-09-01
======================

- Fixed an invalid template escape sequence.
- Updated the default configuration files.
- Fix RPATH handling for private libraries like Duktape.

irccd 3.0.0 2019-08-15
======================

This is a major release. See MIGRATING.md file for more information.

irccd
-----

- New sections `[paths]` and `[paths.plugin]` have been added to control
  standard paths for both irccd and plugins.
- Irccd no longer supports uid, gid, pid and daemon features.
- Sections `[identity]` and `[server]` have been merged.
- Local transports support SSL.
- The origin in rule is now first class value.
- New option `ipv4` in `[transport]`.
- New option `ipv4` in `[server]`.
- Section `[format]` is renamed to `[templates]`.
- New commands are available as irccd arguments `info` and `version`.

irccdctl
--------

- New option `ipv4` in `[connect]`.
- New option `-o` in `rule-add`.
- New option `-o` and `-O` in `rule-edit`.

irccd-test
----------

- A brand new `irccd-test` program has been added to tests plugins on the
  command line.

cmake
-----

- CMake no longer create a fake installation directory while building.
- All targets are placed into the `bin` directory while building.

network API
-----------


plugins/roulette/roulette.js
/*
 * roulette.js -- russian roulette game
 *
 * Copyright (c) 2013-2026 David Demelier <markand@malikania.fr>
 *
 * Permission to use, copy, modify, and/or distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

// Plugin information.
info = {
	author: "David Demelier <markand@malikania.fr>",
	license: "ISC",
	summary: "A russian roulette for IRC",
	version: "@irccd_VERSION@"
};

// Modules.
var Logger = Irccd.Logger;
var Plugin = Irccd.Plugin;
var Server = Irccd.Server;
var Util = Irccd.Util;

/**
 * Formats for writing.
 */
Plugin.templates = {
	"lucky":        "#{nickname}, you're lucky this time",
	"shot":         "HEADSHOT"
};

function Gun(server, channel)
{
	this.server = server;
	this.channel = channel;
	this.index = 0;
	this.bullet = Math.floor(Math.random() * 6);
}

/**
 * Map of games.
 */
Gun.map = {};

/**
 * Search for an existing game.
 *
 * @param server the server object
 * @param channel the channel name
 * @return the hangman instance or undefined if no one exists
 */
Gun.find = function (server, channel)
{
	return Gun.map[server.toString() + '@' + channel];
}

/**
 * Create a new game, store it in the map and return it.
 *
 * @param server the server object
 * @param channel the channel name
 * @return the hangman object
 */
Gun.create = function (server, channel)
{
	return Gun.map[server.toString() + "@" + channel] = new Gun(server, channel);
}

/**
 * Remove the specified game from the map.
 *
 * @param game the game to remove
 */
Gun.remove = function (game)
{
	delete Gun.map[game.server + "@" + game.channel];
}

Gun.prototype.shot = function ()
{
	return this.index++ === this.bullet;
}

function onCommand(server, origin, channel)
{
	channel = channel.toLowerCase();

	var kw = {
		channel: channel,
		command: server.info().commandChar + Plugin.info().name,
		nickname: Util.splituser(origin),
		origin: origin,
		server: server.toString(),
		plugin: Plugin.info().name
	};

	var game = Gun.find(server, channel);

	if (!game)
		game = Gun.create(server, channel);

	if (game.shot()) {
		server.kick(Util.splituser(origin), channel, Util.format(Plugin.templates["shot"], kw));
		Gun.remove(game);
	} else {
		kw.count = (6 - game.index).toString();
		server.message(channel, Util.format(Plugin.templates["lucky"], kw));
	}
}
